libcontainer/rootfs_linux.go

```diff
@@ -152,14 +153,46 @@ func mountToRootfs(m *configs.Mount, rootfs, mountLabel string) error {
 		}
 		return nil
 	case "tmpfs":
+		copyUp := m.Extensions&configs.EXT_COPYUP == configs.EXT_COPYUP
+		tmpDir := ""
 		stat, err := os.Stat(dest)
 		if err != nil {
 			if err := os.MkdirAll(dest, 0755); err != nil {
 				return err
 			}
 		}
-		if err := mountPropagate(m, rootfs, mountLabel); err != nil {
-			return err
+		if copyUp {
+			tmpDirPrefix := os.Getenv("XDG_RUNTIME_DIR")
+			if tmpDirPrefix == "" {
+				tmpDirPrefix = "/tmp"
+			}
+			tmpDir, err = ioutil.TempDir(tmpDirPrefix, "runctmpdir")
+			if err != nil {
+				return fmt.Errorf("failed to create tmpdir: %v", err)
+			}
+			defer os.RemoveAll(tmpDir)
+			m.Destination = tmpDir
+			if err := mountPropagate(m, rootfs, mountLabel); err != nil {
+				return err
+			}
+			if err := fileutils.CopyDirectory(dest, tmpDir); err != nil {
+				errMsg := fmt.Errorf("failed to copy %s to %s: %v", dest, tmpDir, err)
+				if err1 := syscall.Unmount(tmpDir, syscall.MNT_DETACH); err1 != nil {
+					return fmt.Errorf("%v: failed to unmount: %v", errMsg, err1)
+				}
+				return errMsg
+			}
+			if err := syscall.Mount(tmpDir, dest, "", syscall.MS_MOVE, ""); err != nil {
+				errMsg := fmt.Errorf("failed to move mount %s to %s: %v", tmpDir, dest, err)
+				if err1 := syscall.Unmount(tmpDir, syscall.MNT_DETACH); err1 != nil {
+					return fmt.Errorf("%v: failed to unmount: %v", errMsg, err1)
+				}
+				return errMsg
+			}
+		} else {
+			if err := mountPropagate(m, rootfs, mountLabel); err != nil {
+				return err
+			}
```

Comment:
I'd prefer we did something like this, because it would make it more clear what the difference with and without `copyUp` is:

```
if copyUp {
    // change m.Destination to a tmpdir
}

if err := mountPropagate(m, rootfs, mountLabel); err != nil {
    return err
}

if copyUp {
    // copy everything
}
```